bbb/meta-bbb/recipes-core/trebletrouble/files/src/libfft.c

```diff
@@ -0,0 +1,127 @@
+/* libfft.c - fast Fourier transform library
+**
+** Copyright (C) 1989 by Jef Poskanzer.
+**
+** Permission to use, copy, modify, and distribute this software and its
+** documentation for any purpose and without fee is hereby granted, provided
+** that the above copyright notice appear in all copies and that both that
+** copyright notice and this permission notice appear in supporting
+** documentation.  This software is provided "as is" without express or
+** implied warranty.
+*/
+
+#define _GNU_SOURCE
+#include <stdio.h>
+#include <math.h>
+#include "libfft.h"
+#include "tone.h"
+
+#define MAXFFTSIZE 32768
+#define LOG2_MAXFFTSIZE 15
+
+static int bitreverse[MAXFFTSIZE], bits;
+
+/* initfft - initialize for fast Fourier transform
+**
+** b    power of two such that 2**nu = number of samples
+*/
+int initfft( int b )
+    {
+    register int i, j, k;
+
+    bits = b;
+    if ( bits > LOG2_MAXFFTSIZE )
+	{
+	fprintf(
+	    stderr, "%d is too many bits, max is %d\n", bits, LOG2_MAXFFTSIZE );
+	return -1;
+	}
+
+    for ( i = ( 1 << bits ) - 1; i >= 0; --i )
+	{
+	k = 0;
+	for ( j = 0; j < bits; ++j )
+	    {
+	    k *= 2;
+	    if ( i & ( 1 << j ) )
+		k += 1;
+	    }
+	bitreverse[i] = k;
+	}
+    return 0;
+    }
+
+/* fft - a fast Fourier transform routine
+**
+** xr   real part of data to be transformed
+** xi   imaginary part (normally zero, unless inverse transform in effect)
+** inv  flag for inverse
+*/
+
+/*void fft(Wave xr, Wave xi, int inv )*/

```

Comment:
👉 We should wrap the function `fft` so that it takes a single argument of type `Wave*`